topics/products/phpstorm/php_open_api_php_type_providers.md

```diff
@@ -0,0 +1,303 @@
+
+  /**
+   * @param expression to complete - Here you can use index lookups
+   * @param project well so you can reach the PhpIndex based stuff
+   * @return type for element, null if no insight. You can return a custom signature here to be later decoded by getBySignature.
+   */
+  @Nullable
+  PhpType complete(String expression, Project project);
+
+  /**
+   * Here you can extend the signature lookups
+   * @param expression Signature expression to decode. You can use PhpIndex.getBySignature() to look up expression internals.
+   * @param visited Recursion guard: please pass this on into any phpIndex calls having same parameter
+   * @param depth Recursion guard: please pass this on into any phpIndex calls having same parameter
+   * @param project well so you can reach the PhpIndex
+   * @return null if no match
+   */
+  Collection<? extends PhpNamedElement> getBySignature(String expression, Set<String> visited, int depth, Project project);
+
+  default boolean emptyResultIsComplete() {
+    return false;
+  }
+
+  @Internal
+  default boolean interceptsNativeSignature() {
+    return false;
+  }
+
+}
+```
+
+To implement `PhpTypeProvider4`, you need to override 4 methods:
+
+1. `getKey()` is a method that returns a character that will be unique for this type provider.
+   This can be any character, as long as it is unique, for example, PhpStorm uses hieroglyphs.
+   For the convenience of checking that the Incomplete type is a type obtained from your provider,
+   you can use the `PhpCharBasedTypeKey` class in which you need to overload the `getKey()` method that returns a character.
+   In order to check that the type is created by your provider, it will be enough to call the `signed()` method,
+   and in order to sign the type, it will be enough to call the `sign()` method.
+
+2. `getType()` is a method that returns the type of the expression for the given element.
+   This method is called at the indexing stage and therefore its implementation **cannot access** any information
+   from the index and **must** rely only on local information.
+   If you need some information, then pack the required data into a string and return an Incomplete type based on this string.
+
+3. `complete()` is a method that resolves an Incomplete type into a Complete type.
+   All strings of Incomplete types are sequentially passed to this method, the method should return
+   a Complete type for them.
+   Note that Incomplete types from other type providers also come to this method,
+   so you need to check that the incoming string belongs to your provider.
+
+4. `getBySignature()` is a method with which you can provide additional elements or references.
+
+You can also overload the `emptyResultIsComplete()` method, which indicates whether the `null` returned from
+the `complete()` method is a valid result.
+For example, if you index the type `int`, then the result returned from `complete()` will be `null` and it is a valid result.
+
+**Examples:**
```

Comment:
I'd not use there type providers as an idiomatic examples since they're pretty hard to follow and have non-standard implementations. Let's leave `PhpUnitFiledInitializedInSetUpMethodsTP` as the only example implementation

Reply:
Fixed.